File: java/src/linkedlist/LC_19_RemoveNthNodeFromEndOfList.java
package linkedlist;
// https://leetcode.com/problems/remove-nth-node-from-end-of-list

public class LC_19_RemoveNthNodeFromEndOfList {

    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode();
        dummy.next = head;
        ListNode slow = dummy, fast = head;
        for (int i = 0; i < n; i++) {
            fast = fast.next;
        }

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }

        if (slow.next != null)
            slow.next = slow.next.next;

        return dummy.next;
    }
}

File: python/linkedlist/LC_19_RemoveNthNodeFromEndOfList.py
# https://leetcode.com/problems/remove-nth-node-from-end-of-list

from typing import Optional
from linkedlist.listnode import ListNode


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode()
        dummy.next = head
        slow = fast = dummy
        for _ in range(n):
            fast = fast.next

        while fast and fast.next:
            slow = slow.next
            fast = fast.next

        if slow.next:
            slow.next = slow.next.next
        return dummy.next
